src/vehicle.cpp

```diff
@@ -6849,64 +6841,88 @@ long vehicle_part::ammo_capacity() const
 {
     // @todo currently only support fuel tanks and batteries
 
-    if( base.is_magazine() ) {
+    if( base.is_magazine() || base.typeId() == "minireactor" ) {
         return base.ammo_capacity();
     }
 
-    return info().has_flag( VPFLAG_FUEL_TANK ) ? info().size : 0;
+    if( base.is_watertight_container() ) {
+        return base.type->container->contains * std::max( item::find_type( ammo_current() )->stack_size, 1 );
+    }
+
+    return 0;
 }
 
 long vehicle_part::ammo_remaining() const
 {
     // @todo currently only support fuel tanks and batteries
 
-    if( base.is_magazine() ) {
+    if( base.is_magazine() || base.typeId() == "minireactor" ) {
         return base.ammo_remaining();
     }
 
-    return info().has_flag( VPFLAG_FUEL_TANK ) ? amount : 0;
+    if( base.is_watertight_container() ) {
+        return base.contents.empty() ? 0 : base.contents.back().charges;
+    }
+
+    return 0;
 }
 
 int vehicle_part::ammo_set( const itype_id &ammo, long qty )
 {
     // @todo currently only support fuel tanks and batteries
 
-    if( base.is_magazine() ) {
-        base.ammo_set( ammo, qty );
-        return base.ammo_remaining();
-    }
-
-    if( !info().has_flag( VPFLAG_FUEL_TANK ) ) {
+    if( info().fuel_type != ammo ) {
         return -1;
     }
 
-    if( info().fuel_type != ammo ) {
-        return -1;
+    if( qty == 0 ) {
+        ammo_unset();
+        return 0;
     }
 
     if( qty < 0 ) {
         qty = ammo_capacity();
     }
 
-    return amount = std::min( qty, ammo_capacity() );
+    if( base.is_magazine() || base.typeId() == "minireactor" ) {
+        base.ammo_set( ammo, qty );
+        return base.ammo_remaining();
+    }
+
+    if( base.is_watertight_container() ) {
+        base.contents.clear();
+        base.emplace_back( ammo, calendar::turn, std::min( qty, ammo_capacity() ) );
+        return std::min( qty, ammo_capacity() );
+    }
+
+    return -1;
 }
 
 void vehicle_part::ammo_unset() {
-    if( base.is_magazine() ) {
+    if( base.is_magazine() || base.typeId() == "minireactor" ) {
         base.ammo_unset();
-    } else {
-        amount = 0;
+
+    } else if( base.is_watertight_container() ) {
+        base.contents.clear();
     }
 }
 
 long vehicle_part::ammo_consume( long qty, const tripoint& pos )
 {
-    if( base.is_magazine() ) {
+    if( base.is_magazine() || base.typeId() == "minireactor" ) {
         return base.ammo_consume( qty, pos );
     }
 
     int res = std::min( ammo_remaining(), qty );
-    amount -= res;
+
+    if( base.is_watertight_container() ) {
+        item& liquid = base.contents.back();
```

Comment:
`base.contents` can be empty, calling `back` on it is undefined behavior. Thankfully it just crashes on my system when I try to siphon all the fuel out of a tank.